### PR TITLE
[DX-1688, TT-13153] endpoint level rate limits

### PR DESCRIPTION
**Pull request by jeffy-mathew**
### For internal users - Please add a Jira DX PR ticket to the subject!
<br>

---

### Preview Link
https://deploy-preview-5409--tyk-docs.netlify.app/docs/nightly/getting-started/key-concepts/rate-limiting/
https://deploy-preview-5409--tyk-docs.netlify.app/docs/nightly/basic-config-and-security/control-limit-traffic/rate-limiting/

<br>

### Description

<br>

#### Screenshots (if appropriate)
<br>

## Checklist

- [ ] I have added a **preview link** to the PR description.
- [ ] I have reviewed the suggestions made by our AI (PR Agent) and updated them accordingly (spelling errors, rephrasing, etc.)
- [ ] I have [reviewed the guidelines](https://github.com/TykTechnologies/tyk-docs/blob/master/CONTRIBUTING.md) for contributing to this repository.
- [ ] I have [read the technical guidelines](https://github.com/TykTechnologies/tyk-docs/blob/master/CONTRIBUTING-TECHNICAL-GUIDE.md) for contributing to this repository.
- [ ] Make sure you have started *your change* off *our latest `master`*.
- [ ] I **labeled** the PR

___

### **PR Type**
documentation

___

### **Description**
- Added comprehensive documentation for configuring key-level per-endpoint rate limits in Tyk Dashboard.
- Introduced the concept of key-level per-endpoint rate limits in the key concepts section.
- Updated the guidance on selecting appropriate rate limiting levels based on user needs.
- Revised the order of assessment when multiple rate limits are configured to include per-endpoint limits.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>rate-limiting.md</strong><dd><code>Add documentation for key-level per-endpoint rate limits</code>&nbsp; </dd></summary>
<hr>

tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md

<li>Added section on setting up key-level per-endpoint rate limits.<br> <li> Provided detailed steps for configuring endpoint-level rate limits in <br>Tyk Dashboard.<br> <li> Included notes on assumptions for endpoint-level rate limits.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk-docs/pull/5409/files#diff-1d63c4783ce4de0d1fbaed0b2a6e1ed70cf4a5c0557f1790439e294463127572">+21/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>rate-limiting.md</strong><dd><code>Update key concepts with per-endpoint rate limiting</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tyk-docs/content/getting-started/key-concepts/rate-limiting.md

<li>Introduced key-level per-endpoint limit concept.<br> <li> Updated guidance on choosing rate limiting levels.<br> <li> Revised order of assessment for multiple rate limits.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk-docs/pull/5409/files#diff-a0f383800801ae765ec875c3c98610f657751ec0779b9e23970357b74fb58e9a">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md [hunk 1]
```diff
@@ -75,6 +75,26 @@ Check out the following video to see this being done.
 
 {{< youtube n7jbmuWgPsw >}}
 
+## Setting up a Key-Level Per-endpoint rate limit
+
+If you want to restrict API clients to a certain rate of requests for a specific endpoint of an API, you will also configure the rate limiter via the security policy. However this time you'll assign per-endpoint limits. The allowance that you configure in the policy will be consumed by any requests made to that specific endpoint using a key generated from that policy. Thus, if a policy grants access to an API endpoint with `rate=5 per=60` then three clients using keys generated from that policy will each independently be able to make 5 requests in any 60 second period to that particular endpont of the API before receiving the `HTTP 429 Too Many Requests` error. Endpoints of the API that doesn't have endpoint level rate limits configured would be using the rate limits inherited to the API either via per-API rate limit or key level global rate limits based on the configurations.
```
Comment (andyo-tyk) [inline tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md]:
```suggestion
If you want to restrict the request rate for specific API clients to a specific endpoint of an API, you should use the security policy to assign per-endpoint rate limits. The allowance that you configure in the policy will be consumed by any requests made to that specific endpoint using a key generated from that policy. Each key will have its own allowance, for example: if a policy grants access to an API endpoint with `rate=5 per=60` then three clients using keys generated from that policy will each independently be able to make 5 requests in any 60 second period to that particular endpont of the API before receiving the `HTTP 429 Too Many Requests` error.

Any API endpoints that don't have endpoint-level rate limits configured will inherit the key-level per-API or key-level global rate limit depending upon which is configured.
```

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md [hunk 2]
```diff
@@ -75,6 +75,26 @@ Check out the following video to see this being done.
...
+
+{{< note success >}}
+**Note**  
+ Following assumptions are made for the aforementioned setup.
```
Comment (andyo-tyk) [inline tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md]:
```suggestion
The following assumptions are made:
```

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md [hunk 3]
```diff
@@ -75,6 +75,26 @@ Check out the following video to see this being done.
...
+ - The APIs being protected with a rate limit are using our [Authentication token]({{< ref "basic-config-and-security/security/authentication-authorization/bearer-tokens" >}}) Authentication mode and have policies already created.
```
Comment (andyo-tyk) [inline tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md]:
```suggestion
 - the APIs for which rate limits are being set are protected using [authentication token]({{< ref "basic-config-and-security/security/authentication-authorization/bearer-tokens" >}}) and have policies already created
```

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md [hunk 4]
```diff
@@ -75,6 +75,26 @@ Check out the following video to see this being done.
...
+
+{{< note success >}}
+**Note**  
+ Following assumptions are made for the aforementioned setup.
+ - The APIs being protected with a rate limit are using our [Authentication token]({{< ref "basic-config-and-security/security/authentication-authorization/bearer-tokens" >}}) Authentication mode and have policies already created. 
+ - Ignore authentication plugin is not turned on for these endpoints.
```
Comment (andyo-tyk) [inline tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md]:
```suggestion
 - the [ignore authentication]({{< ref "product-stack/tyk-gateway/middleware/ignore-middleware" >}}) middleware is not turned enabled for these endpoints
```

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md [hunk 5]
```diff
@@ -75,6 +75,26 @@ Check out the following video to see this being done.
...
+ - Path based permissions, if configured grants access to these endpoints
```
Comment (andyo-tyk) [inline tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md]:
```suggestion
 - if [path based permissions]({{< ref "getting-started/create-security-policy#path-based-permissions" >}}) are configured for the API then these grant access to these endpoints for keys generated from the policies
```

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md [hunk 6]
```diff
@@ -75,6 +75,26 @@ Check out the following video to see this being done.
...
+{{< /note >}}
+
+You can configure this rate limit from the API Designer in Tyk Dashboard as follows:
```
Comment (andyo-tyk) [inline tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md]:
```suggestion
You can configure per-endpoint rate limits from the API Designer in Tyk Dashboard as follows:
```

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md [hunk 7]
```diff
@@ -75,6 +75,26 @@ Check out the following video to see this being done.
...
+
+1. Navigate to the Tyk policy for which you want to set the rate limit
+2. Ensure that API that you want to apply rate limits to is selected
+3. Under **API Access** -> **Set endpoint-level usage limits** click on **Add Rate Lmit** to configure endpoint level rate limits.
```
Comment (andyo-tyk) [inline tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md]:
```suggestion
3. Under **API Access** -> **Set endpoint-level usage limits** click on **Add Rate Limit** to configure the rate limit. You will need to provide the rate limit and the endpoint path and method.
```

File: tyk-docs/content/getting-started/key-concepts/rate-limiting.md [hunk 8]
```diff
@@ -51,6 +51,7 @@ Key-level rate limiting is more focused on controlling traffic from individual s
 
 - **key-level global limit** limiting the rate of calls the user of a key can make to all APIs authorized by that key
 - **key-level per-API limit** limiting the rate of calls the user of a key can make to specific individual APIs
+- **key-level per-endpoint limit** limiting the rate of calls the user of key can make to specific individual endpoints of an API
```
Comment (andyo-tyk) [inline tyk-docs/content/getting-started/key-concepts/rate-limiting.md]:
```suggestion
- **key-level per-endpoint limit** limiting the rate of calls the user of a key can make to specific individual endpoints of an API
```

File: tyk-docs/content/getting-started/key-concepts/rate-limiting.md [hunk 9]
```diff
@@ -61,6 +62,7 @@ The simplest way to figure out which level of rate limiting you’d like to appl
 - do you want to protect your service against denial of service attacks or overwhelming amounts of traffic from **all users** of the API? **You’ll want to use an API-level rate limit!**
 - do you want to limit the number of requests a specific user can make to **all APIs** they have access to? **You’ll want to use a key-level global rate limit!**
 - do you want to limit the number of requests a specific user can make to **specific APIs** they have access to? **You’ll want to use a key-level per-API rate limit.**
+- do you want to limit the number of requests a specific user can make to **specific endpoint of an API** they have access to? **You’ll want to use a key-level per-endpoint rate limit.**
```
Comment (andyo-tyk) [inline tyk-docs/content/getting-started/key-concepts/rate-limiting.md]:
```suggestion
- do you want to limit the number of requests a specific user can make to a **specific endpoint of an API** they have access to? **You’ll want to use a key-level per-endpoint rate limit.**
```

**Review by andyo-tyk — CHANGES_REQUESTED** [review]
Some minor suggestions

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md (hunk 3 above)
Comment (andyo-tyk) [reply tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md]: This assumption is not valid and has been removed.

**Review by andyo-tyk — APPROVED** [review]
LGTM, thanks.

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md [hunk 10]
```diff
@@ -75,6 +75,27 @@ Check out the following video to see this being done.
 
 {{< youtube n7jbmuWgPsw >}}
 
+## Setting up a Key-Level Per-endpoint rate limit
+
+If you want to restrict the request rate for specific API clients to a specific endpoint of an API, you should use the security policy to assign per-endpoint rate limits. The allowance that you configure in the policy will be consumed by any requests made to that specific endpoint using a key generated from that policy. Each key will have its own allowance, for example: if a policy grants access to an API endpoint with `rate=5 per=60` then three clients using keys generated from that policy will each independently be able to make 5 requests in any 60 second period to that particular endpont of the API before receiving the `HTTP 429 Too Many Requests` error.
+
+Any API endpoints that don't have endpoint-level rate limits configured will inherit the key-level per-API or key-level global rate limit depending upon which is configured.
```
Comment (sharadregoti) [inline tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md]:
```suggestion
To restrict the request rate for specific API clients on particular endpoints, you can use the security policy to assign per-endpoint rate limits. These limits are set within the policy and will be #enforced for any requests made to that endpoint by clients using keys generated from that policy.

Each key will have its own independent rate limit allowance. For example, if a policy grants access to an endpoint with a rate limit of 5 requests per 60 seconds, each client with a key from that policy can make 5 requests to the endpoint in any 60-second period. Once the limit is reached, the client will receive an HTTP 429 Too Many Requests error.

If no per-endpoint rate limit is defined, the endpoint will inherit the key-level per-API rate limit or the global rate limit, depending on what is configured.
```

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md [hunk 11]
```diff
@@ -75,6 +75,27 @@ Check out the following video to see this being done.
...
+
+{{< note success >}}
+**Note**  
+The following assumptions are made:
+ - the [ignore authentication]({{< ref "product-stack/tyk-gateway/middleware/ignore-middleware" >}}) middleware is not turned enabled for these endpoints
+ - if [path based permissions]({{< ref "getting-started/create-security-policy#path-based-permissions" >}}) are configured for the API then these grant access to these endpoints for keys generated from the policies
```
Comment (sharadregoti) [inline tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md]:
```suggestion
 - If [path-based permissions]({{< ref "getting-started/create-security-policy#path-based-permissions" >}}) are configured, they must grant access to these endpoints for keys generated from the policies. {{< /note >}}
```

**Comment by letzya** [comment]
@jeffy-mathew @andyo-tyk going forward, please check the **PR Code Suggestions** bot. It found a typo etc.

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md [hunk 12]
```diff
@@ -75,6 +75,27 @@ Check out the following video to see this being done.
 
 {{< youtube n7jbmuWgPsw >}}
 
+## Setting up a Key-Level Per-endpoint rate limit
+
+If you want to restrict the request rate for specific API clients to a specific endpoint of an API, you should use the security policy to assign per-endpoint rate limits. The allowance that you configure in the policy will be consumed by any requests made to that specific endpoint using a key generated from that policy. Each key will have its own allowance, for example: if a policy grants access to an API endpoint with `rate=5 per=60` then three clients using keys generated from that policy will each independently be able to make 5 requests in any 60 second period to that particular endpont of the API before receiving the `HTTP 429 Too Many Requests` error.
```
Comment (letzya) [inline tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md]:
```suggestion
If you want to restrict the request rate for specific API clients to a specific endpoint of an API, you should use the security policy to assign per-endpoint rate limits. The allowance that you configure in the policy will be consumed by any requests made to that specific endpoint using a key generated from that policy. Each key will have its own allowance, for example: if a policy grants access to an API endpoint with `rate=5 per=60` then three clients using keys generated from that policy will each independently be able to make 5 requests in any 60 second period to that particular endpoint of the API before receiving the `HTTP 429 Too Many Requests` error.
```

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md [hunk 13]
```diff
@@ -75,6 +75,27 @@ Check out the following video to see this being done.
...
+## Setting up a Key-Level Per-endpoint rate limit
+
+If you want to restrict the request rate for specific API clients to a specific endpoint of an API, you should use the security policy to assign per-endpoint rate limits. The allowance that you configure in the policy will be consumed by any requests made to that specific endpoint using a key generated from that policy. Each key will have its own allowance, for example: if a policy grants access to an API endpoint with `rate=5 per=60` then three clients using keys generated from that policy will each independently be able to make 5 requests in any 60 second period to that particular endpoint of the API before receiving the `HTTP 429 Too Many Requests` error.
+
+Any API endpoints that don't have endpoint-level rate limits configured will inherit the key-level per-API or key-level global rate limit depending upon which is configured.
+
+{{< note success >}}
+**Note**  
+The following assumptions are made:
+ - the [ignore authentication]({{< ref "product-stack/tyk-gateway/middleware/ignore-middleware" >}}) middleware is not turned enabled for these endpoints
```
Comment (letzya) [inline tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md]:
```suggestion
 - the [ignore authentication middleware]({{< ref "product-stack/tyk-gateway/middleware/ignore-middleware" >}}) is not enabled for these endpoints
```

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md [hunk 14]
```diff
@@ -75,6 +75,27 @@ Check out the following video to see this being done.
 
 {{< youtube n7jbmuWgPsw >}}
 
+## Setting up a Key-Level Per-endpoint rate limit
```
Comment (letzya) [inline tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md]:
```suggestion
## Setting up a key-level per-endpoint rate limit
```

**Review by letzya — DISMISSED** [review]
Maybe a couple of suggestions + check my slack message @jeffy-mathew

File: tyk-docs/content/getting-started/key-concepts/rate-limiting.md [hunk 15]
```diff
@@ -69,6 +71,7 @@ When multiple rate limits are configured, they are assessed in this order (if ap
 1. API-level global rate limit
 2. Key-level global rate limit
 3. Key-level per-API rate limit
+4. Key-level per-endpoint rate limit
```
Comment (letzya) [inline tyk-docs/content/getting-started/key-concepts/rate-limiting.md]: @jeffy-mathew we also have `API-level per-endpoint rate limit`

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md [hunk 16]
```diff
@@ -75,6 +75,27 @@ Check out the following video to see this being done.
 
 {{< youtube n7jbmuWgPsw >}}
 
+## Setting up a key-level per-endpoint rate limit
+
+If you want to restrict the request rate for specific API clients to a specific endpoint of an API, you should use the security policy to assign per-endpoint rate limits. The allowance that you configure in the policy will be consumed by any requests made to that specific endpoint using a key generated from that policy. Each key will have its own allowance, for example: if a policy grants access to an API endpoint with `rate=5 per=60` then three clients using keys generated from that policy will each independently be able to make 5 requests in any 60 second period to that particular endpoint of the API before receiving the `HTTP 429 Too Many Requests` error.
+
+Any API endpoints that don't have endpoint-level rate limits configured will inherit the key-level per-API or key-level global rate limit depending upon which is configured.
```
Comment (sharadregoti) [inline tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md:82]:
```suggestion
To restrict the request rate for specific API clients on particular endpoints, you can use the security policy to assign per-endpoint rate limits. These limits are set within the policy and will be #enforced for any requests made to that endpoint by clients using keys generated from that policy.
Each key will have its own independent rate limit allowance. For example, if a policy grants access to an endpoint with a rate limit of 5 requests per 60 seconds, each client with a key from that policy can make 5 requests to the endpoint in any 60-second period. Once the limit is reached, the client will receive an HTTP 429 Too Many Requests error.
If no per-endpoint rate limit is defined, the endpoint will inherit the key-level per-API rate limit or the global rate limit, depending on what is configured.
```

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md [hunk 17]
```diff
@@ -75,6 +75,27 @@ Check out the following video to see this being done.
...
+
+{{< note success >}}
+**Note**  
+The following assumptions are made:
+ - the [ignore authentication middleware]({{< ref "product-stack/tyk-gateway/middleware/ignore-middleware" >}}) is not enabled for these endpoints
```
Comment (sharadregoti) [inline tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md]:
```suggestion
 - The [ignore authentication]({{< ref "product-stack/tyk-gateway/middleware/ignore-middleware" >}}) middleware should not be enabled for the relevant endpoints.
```

**Review by sharadregoti — CHANGES_REQUESTED** [review]

File: tyk-docs/content/getting-started/key-concepts/rate-limiting.md [hunk 18]
```diff
@@ -69,6 +71,7 @@ When multiple rate limits are configured, they are assessed in this order (if ap
 1. API-level global rate limit
```
Comment (andyo-tyk) [inline tyk-docs/content/getting-started/key-concepts/rate-limiting.md]:
```suggestion
1. API-level per-endpoint rate limit (configured in API definition)
2. API-level rate limit (configured in API definition)
```

File: tyk-docs/content/getting-started/key-concepts/rate-limiting.md [hunk 19]
```diff
@@ -69,6 +71,7 @@ When multiple rate limits are configured, they are assessed in this order (if ap
...
 2. Key-level global rate limit
```
Comment (andyo-tyk) [inline tyk-docs/content/getting-started/key-concepts/rate-limiting.md]:
```suggestion
3. Key-level per-endpoint rate limit (configured in access key)
```

File: tyk-docs/content/getting-started/key-concepts/rate-limiting.md [hunk 20]
```diff
@@ -69,6 +71,7 @@ When multiple rate limits are configured, they are assessed in this order (if ap
...
 3. Key-level per-API rate limit
```
Comment (andyo-tyk) [inline tyk-docs/content/getting-started/key-concepts/rate-limiting.md]:
```suggestion
4. Key-level per-API rate limit (configured in access key)
```

File: tyk-docs/content/getting-started/key-concepts/rate-limiting.md (hunk 15 above)
Comment (andyo-tyk) [inline tyk-docs/content/getting-started/key-concepts/rate-limiting.md]:
```suggestion
5. Key-level global rate limit (configured in access key)
```

**Review by andyo-tyk — CHANGES_REQUESTED** [review]

**Review by andyo-tyk — APPROVED** [review]

**Review by andyo-tyk — APPROVED** [review]

File: tyk-docs/content/getting-started/key-concepts/rate-limiting.md [hunk 21]
```diff
@@ -61,14 +62,17 @@ The simplest way to figure out which level of rate limiting you’d like to appl
 - do you want to protect your service against denial of service attacks or overwhelming amounts of traffic from **all users** of the API? **You’ll want to use an API-level rate limit!**
```
Comment (sharadregoti) [inline tyk-docs/content/getting-started/key-concepts/rate-limiting.md:62]: Should we also add an example for `API-level per-endpoint rate limit (configured in API definition)`

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md (hunk 10 above)
Comment (sharadregoti) [reply tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md]: @andyo-tyk I think you missed this rephrasing suggestion, do you want to include it?

**Review by sharadregoti — CHANGES_REQUESTED** [review]

File: tyk-docs/content/getting-started/key-concepts/rate-limiting.md [hunk 22]
```diff
@@ -61,14 +62,17 @@ The simplest way to figure out which level of rate limiting you’d like to appl
...
 - do you want to limit the number of requests a specific user can make to **all APIs** they have access to? **You’ll want to use a key-level global rate limit!**
```
Comment (andyo-tyk) [inline tyk-docs/content/getting-started/key-concepts/rate-limiting.md:64]:
```suggestion
- do you have a health endpoint that consumes very little resource on your service and can handle significantly more requests than your other endpoints? **You'll want to use an API-level per-endpoint rate limit!**
- do you want to limit the number of requests a specific user can make to **all APIs** they have access to? **You’ll want to use a key-level global rate limit!**
```

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md (hunk 10 above)
Comment (andyo-tyk) [reply tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md]: Yes, looks good to me (though we usually put the HTTP response codes/messages in `code` format.

**Review by sharadregoti — APPROVED** [review]

**Pull request merged**